Backend/Remora.Discord.Caching.Redis/Services/RedisCacheService.cs

```diff
@@ -0,0 +1,353 @@
+    }
+
+    /// <inheritdoc />
+    public async ValueTask CacheAsync<TInstance>(string key, TInstance instance) where TInstance : class
+    {
+        if (_cacheSettings.GetAbsoluteExpirationOrDefault(typeof(TInstance)) is var absoluteExpiration)
+        {
+            if (absoluteExpiration == TimeSpan.Zero)
+            {
+                return;
+            }
+        }
+
+        Task cacheTask = instance switch
+        {
+            IWebhook webhook => CacheWebhook(key, webhook),
+            ITemplate template => CacheTemplate(key, template),
+            IIntegration integration => CacheIntegration(key, integration),
+            IBan ban => CacheBan(key, ban),
+            IGuildMember member => CacheGuildMember(key, member),
+            IGuildPreview preview => CacheGuildPreview(key, preview),
+            IGuild guild => CacheGuild(key, guild),
+            IEmoji emoji => CacheEmoji(key, emoji),
+            IInvite invite => CacheInvite(key, invite),
+            IMessage message => CacheMessage(key, message),
+            IChannel channel => CacheChannel(key, channel),
+            _ => CacheInstanceAsync(key, instance)
+        };
+
+        await cacheTask;
+    }
+
+    /// <inheritdoc />
+    public async ValueTask<Result<TInstance>> TryGetValueAsync<TInstance>(string key) where TInstance : class
+    {
+        var cacheResult = await _cache.GetAsync(key);
+
+        if (cacheResult is null)
+        {
+            return Result<TInstance>.FromError(new NotFoundError($"The key \"{key}\" was not found in the cache."));
+        }
+
+        var deserialized = JsonSerializer.Deserialize<TInstance>(cacheResult, _jsonOptions)!;
+
+        await _cache.RefreshAsync(key);
+
+        return Result<TInstance>.FromSuccess(deserialized);
+    }
+
+    /// <inheritdoc />
+    public async ValueTask<Result<TInstance>> TryGetPreviousValueAsync<TInstance>(string key) where TInstance : class
+    {
+        var evictionKey = $"Evicted:{key}";
+
+        var cacheResult = await _cache.GetAsync(evictionKey);
+
+        if (cacheResult is null)
+        {
+            return Result<TInstance>.FromError(new NotFoundError($"The key \"{key}\" was not found in the eviction cache."));
+        }
+
+        var deserialized = JsonSerializer.Deserialize<TInstance>(cacheResult, _jsonOptions)!;
+
+        await _cache.RefreshAsync(key);
+
+        return Result<TInstance>.FromSuccess(deserialized);
+    }
+
+    /// <inheritdoc />
+    public async ValueTask EvictAsync<TInstance>(string key) where TInstance : class
+    {
+        var value = await _cache.GetAsync(key);
+
+        if (value is null)
+        {
+            return;
+        }
+
+        var evictionKey = $"Evicted:{key}";
+        var settings = _cacheSettings.GetRedisEntryOptions<TInstance>();
+
+        await _cache.SetAsync(evictionKey, value, settings);
+    }
+
+    private async Task CacheWebhook(string key, IWebhook webhook)
```

Comment:
Could we break these out into a shared implementation that uses the cache service under the hood? It's a little silly to have exactly the same code in two places.